flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java

```diff
@@ -0,0 +1,416 @@
+
+		if (readers.isEmpty()) {
+			memory.unmapAndDispose();
+		}
+	}
+
+	// ------------------------------ legacy ----------------------------------
+
+	@Override
+	public int releaseMemory() throws IOException {
+		return 0;
+	}
+
+	// ---------------------------- statistics --------------------------------
+
+	@Override
+	public int unsynchronizedGetNumberOfQueuedBuffers() {
+		return 0;
+	}
+
+	@Override
+	protected long getTotalNumberOfBuffers() {
+		return numBuffersAndEventsWritten;
+	}
+
+	@Override
+	protected long getTotalNumberOfBytes() {
+		return memory.getSize();
+	}
+
+	int getBuffersInBacklog() {
+		return numDataBuffersWritten;
+	}
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * This class is largely a workaround for the fact that a memory mapped region in Java can cannot
+	 * be larger than 2GB (== signed 32 bit int max value).
+	 *
+	 * <p>The class takes {@link Buffer}s and writes them to several memory mapped region, using the
+	 * {@link BoundedBlockingSubpartitionMemory} class.
+	 *
+	 * <p>This class performs absolutely no synchronization and relies on single threaded
+	 * or externally synchronized access.
+	 */
+	private static final class MemoryMappedBuffers {
+
+		/** Memory mappings should be at the granularity of page sizes, for efficiency. */
+		private static final int PAGE_SIZE = PageSizeUtil.getSystemPageSizeOrConservativeMultiple();
+
+		/** The encoder to the current memory mapped region we are writing to. */
+		private BoundedBlockingSubpartitionMemory.Writer currentBuffer;
+
+		/** All memory mapped regions that are already full (completed). */
+		private final ArrayList<ByteBuffer> fullBuffers;
+
+		/** The file channel backing the memory mapped file. */
+		private final FileChannel file;
+
+		/** The path of the memory mapped file. */
+		private final Path filePath;
+
+		/** The offset where the next mapped region should start. */
+		private long nextMappingOffset;
+
+		/** The size of each mapped region. */
+		private final long mappingSize;
+
+		MemoryMappedBuffers(Path filePath, FileChannel file, int maxSizePerByteBuffer) throws IOException {
+			this.filePath = filePath;
+			this.file = file;
+			this.mappingSize = alignSize(maxSizePerByteBuffer);
+			this.fullBuffers = new ArrayList<>(4);
+
+			rollOverToNextBuffer();
+		}
+
+		void writeBuffer(Buffer buffer) throws IOException {
+			if (currentBuffer.writeBuffer(buffer)) {
+				return;
+			}
+
+			rollOverToNextBuffer();
+
+			if (!currentBuffer.writeBuffer(buffer)) {
+				throwTooLargeBuffer(buffer);
+			}
+		}
+
+		List<ByteBuffer> getFullBuffers() {
+			return fullBuffers.stream()
+					.map(ByteBuffer::slice)
+					.collect(Collectors.toList());
+		}
+
+		/**
+		 * Finishes the current region and prevents further writes.
+		 */
+		void finishWrite() throws IOException {
+			fullBuffers.add(currentBuffer.complete());
+			currentBuffer = null; // fail further writes fast
```

Comment:
Because that class is a private inner class, I assume a collaboration on the contract with the outer class. The outer class has checks in place that this situation cannot occur.